File: lang/ts-plugins/src/hierarchy/entry-exit-compiler.ts
import type { HierarchicalFlowSpec, HierarchicalStateSpec, HierarchicalTransitionSpec } from './types.js';

export class EntryExitCompiler {
  synthesize(spec: HierarchicalFlowSpec): HierarchicalTransitionSpec[] {
    const generated: HierarchicalTransitionSpec[] = [];
    for (const state of spec.rootStates) {
      this.walkEntryExit(state, generated, null);
    }
    return generated;
  }

  compileTransition(spec: HierarchicalFlowSpec, from: string, to: string): HierarchicalTransitionSpec[] {
    const pathFrom = this.pathTo(spec.rootStates, from);
    const pathTo = this.pathTo(spec.rootStates, to);
    if (pathFrom.length === 0 || pathTo.length === 0) return [];

    const lcaIdx = this.lcaIndex(pathFrom, pathTo);
    const result: HierarchicalTransitionSpec[] = [];

    for (let i = pathFrom.length - 1; i > lcaIdx; i--) {
      const state = pathFrom[i];
      if (state.exitProduces.length > 0) {
        result.push({
          from: state.name,
          to: `${state.name}__EXIT_END`,
          trigger: `__exit__${state.name}`,
          requires: [],
          produces: [...state.exitProduces],
        });
      }
    }

    for (let i = lcaIdx + 1; i < pathTo.length; i++) {
      const state = pathTo[i];
      if (state.entryProduces.length > 0) {
        const parentName = pathTo[i - 1].name;
        result.push({
          from: parentName,
          to: state.name,
          trigger: `__entry__${state.name}`,
          requires: [],
          produces: [...state.entryProduces],
        });
      }
    }

    return result;
  }

  synthesizeBubbling(spec: HierarchicalFlowSpec): HierarchicalTransitionSpec[] {
    const handledTriggers = new Map<string, Set<string>>();
    for (const t of spec.transitions) {
      if (!handledTriggers.has(t.from)) handledTriggers.set(t.from, new Set());
      handledTriggers.get(t.from)!.add(t.trigger);
    }

    const bubbled: HierarchicalTransitionSpec[] = [];
    for (const t of spec.transitions) {
      this.propagateBubbling(spec.rootStates, t, handledTriggers, bubbled);
    }
    return bubbled;
  }

  lca(spec: HierarchicalFlowSpec, a: string, b: string): HierarchicalStateSpec | null {
    const pathA = this.pathTo(spec.rootStates, a);
    const pathB = this.pathTo(spec.rootStates, b);
    const idx = this.lcaIndex(pathA, pathB);
    return idx >= 0 ? pathA[idx] : null;
  }

  pathTo(roots: HierarchicalStateSpec[], name: string): HierarchicalStateSpec[] {
    for (const root of roots) {
      const path: HierarchicalStateSpec[] = [];
      if (this.findPath(root, name, path)) return path;
    }
    return [];
  }

  private findPath(node: HierarchicalStateSpec, name: string, path: HierarchicalStateSpec[]): boolean {
    path.push(node);
    if (node.name === name) return true;
    for (const child of node.children) {
      if (this.findPath(child, name, path)) return true;
    }
    path.pop();
    return false;
  }

  private lcaIndex(pathA: HierarchicalStateSpec[], pathB: HierarchicalStateSpec[]): number {
    let lca = -1;
    for (let i = 0; i < Math.min(pathA.length, pathB.length); i++) {
      if (pathA[i].name === pathB[i].name) {
        lca = i;
      } else {
        break;
      }
    }
    return lca;
  }

  private walkEntryExit(
    state: HierarchicalStateSpec,
    out: HierarchicalTransitionSpec[],
    parentName: string | null,
  ): void {
    if (state.entryProduces.length > 0) {
      out.push({
        from: parentName ?? `${state.name}__ENTRY_START`,
        to: state.name,
        trigger: `__entry__${state.name}`,
        requires: [],
        produces: [...state.entryProduces],
      });
    }
    if (state.exitProduces.length > 0) {
      out.push({
        from: state.name,
        to: `${state.name}__EXIT_END`,
        trigger: `__exit__${state.name}`,
        requires: [],
        produces: [...state.exitProduces],
      });
    }
    for (const child of state.children) {
      this.walkEntryExit(child, out, state.name);
    }
  }

  private propagateBubbling(
    roots: HierarchicalStateSpec[],
    parentTransition: HierarchicalTransitionSpec,
    handledTriggers: Map<string, Set<string>>,
    out: HierarchicalTransitionSpec[],
  ): void {
    const fromState = this.findState(roots, parentTransition.from);
    if (!fromState) return;

    for (const child of fromState.children) {
      const childHandles = handledTriggers.get(child.name) ?? new Set();
      if (!childHandles.has(parentTransition.trigger)) {
        const fallback: HierarchicalTransitionSpec = {
          from: child.name,
          to: parentTransition.to,
          trigger: parentTransition.trigger,
          requires: [...parentTransition.requires],
          produces: [...parentTransition.produces],
        };
        out.push(fallback);
        this.propagateBubbling(roots, fallback, handledTriggers, out);
      }
    }
  }

  private findState(roots: HierarchicalStateSpec[], name: string): HierarchicalStateSpec | null {
    for (const root of roots) {
      const found = this.findStateRecursive(root, name);
      if (found) return found;
    }
    return null;
  }

  private findStateRecursive(node: HierarchicalStateSpec, name: string): HierarchicalStateSpec | null {
    if (node.name === name) return node;
    for (const child of node.children) {
      const found = this.findStateRecursive(child, name);
      if (found) return found;
    }
    return null;
  }
}

File: lang/java-plugins/src/main/java/org/unlaxer/tramli/plugins/hierarchy/EntryExitCompiler.java
package org.unlaxer.tramli.plugins.hierarchy;

import java.util.*;

public final class EntryExitCompiler {

    public List<HierarchicalTransitionSpec> synthesize(HierarchicalFlowSpec spec) {
        List<HierarchicalTransitionSpec> generated = new ArrayList<>();
        for (HierarchicalStateSpec state : spec.rootStates()) {
            synthesizeEntryExit(state, generated, null);
        }
        return generated;
    }

    public List<HierarchicalTransitionSpec> compileTransition(HierarchicalFlowSpec spec, String from, String to) {
        List<HierarchicalStateSpec> pathFrom = pathTo(spec.rootStates(), from);
        List<HierarchicalStateSpec> pathTo = pathTo(spec.rootStates(), to);
        if (pathFrom.isEmpty() || pathTo.isEmpty()) {
            return List.of();
        }

        int lcaIndex = lcaIndex(pathFrom, pathTo);
        List<HierarchicalTransitionSpec> result = new ArrayList<>();

        for (int i = pathFrom.size() - 1; i > lcaIndex; i--) {
            HierarchicalStateSpec state = pathFrom.get(i);
            if (!state.exitProduces().isEmpty()) {
                HierarchicalTransitionSpec exit = new HierarchicalTransitionSpec(
                        state.name(), state.name() + "__EXIT_END", "__exit__" + state.name());
                state.exitProduces().forEach(exit::produces);
                result.add(exit);
            }
        }

        for (int i = lcaIndex + 1; i < pathTo.size(); i++) {
            HierarchicalStateSpec state = pathTo.get(i);
            if (!state.entryProduces().isEmpty()) {
                String parentName = pathTo.get(i - 1).name();
                HierarchicalTransitionSpec entry = new HierarchicalTransitionSpec(
                        parentName, state.name(), "__entry__" + state.name());
                state.entryProduces().forEach(entry::produces);
                result.add(entry);
            }
        }

        return result;
    }

    public List<HierarchicalTransitionSpec> synthesizeBubbling(HierarchicalFlowSpec spec) {
        Map<String, Set<String>> handledTriggers = new HashMap<>();
        for (HierarchicalTransitionSpec t : spec.transitions()) {
            handledTriggers.computeIfAbsent(t.from(), k -> new HashSet<>()).add(t.trigger());
        }

        List<HierarchicalTransitionSpec> bubbled = new ArrayList<>();
        for (HierarchicalTransitionSpec t : spec.transitions()) {
            propagateBubbling(spec.rootStates(), t, handledTriggers, bubbled);
        }
        return bubbled;
    }

    public HierarchicalStateSpec lca(HierarchicalFlowSpec spec, String a, String b) {
        List<HierarchicalStateSpec> pathA = pathTo(spec.rootStates(), a);
        List<HierarchicalStateSpec> pathB = pathTo(spec.rootStates(), b);
        int idx = lcaIndex(pathA, pathB);
        return idx >= 0 ? pathA.get(idx) : null;
    }

    List<HierarchicalStateSpec> pathTo(List<HierarchicalStateSpec> roots, String name) {
        for (HierarchicalStateSpec root : roots) {
            List<HierarchicalStateSpec> path = new ArrayList<>();
            if (findPath(root, name, path)) {
                return path;
            }
        }
        return List.of();
    }

    private boolean findPath(HierarchicalStateSpec node, String name, List<HierarchicalStateSpec> path) {
        path.add(node);
        if (node.name().equals(name)) return true;
        for (HierarchicalStateSpec child : node.children()) {
            if (findPath(child, name, path)) return true;
        }
        path.remove(path.size() - 1);
        return false;
    }

    private int lcaIndex(List<HierarchicalStateSpec> pathA, List<HierarchicalStateSpec> pathB) {
        int lca = -1;
        for (int i = 0; i < Math.min(pathA.size(), pathB.size()); i++) {
            if (pathA.get(i).name().equals(pathB.get(i).name())) {
                lca = i;
            } else {
                break;
            }
        }
        return lca;
    }

    private void synthesizeEntryExit(HierarchicalStateSpec state, List<HierarchicalTransitionSpec> out, String parentName) {
        if (!state.entryProduces().isEmpty()) {
            HierarchicalTransitionSpec entry = new HierarchicalTransitionSpec(
                    parentName == null ? state.name() + "__ENTRY_START" : parentName,
                    state.name(),
                    "__entry__" + state.name());
            state.entryProduces().forEach(entry::produces);
            out.add(entry);
        }
        if (!state.exitProduces().isEmpty()) {
            HierarchicalTransitionSpec exit = new HierarchicalTransitionSpec(
                    state.name(),
                    state.name() + "__EXIT_END",
                    "__exit__" + state.name());
            state.exitProduces().forEach(exit::produces);
            out.add(exit);
        }
        for (HierarchicalStateSpec child : state.children()) {
            synthesizeEntryExit(child, out, state.name());
        }
    }

    private void propagateBubbling(List<HierarchicalStateSpec> roots, HierarchicalTransitionSpec parentTransition,
                                   Map<String, Set<String>> handledTriggers, List<HierarchicalTransitionSpec> out) {
        HierarchicalStateSpec fromState = findState(roots, parentTransition.from());
        if (fromState == null) return;

        for (HierarchicalStateSpec child : fromState.children()) {
            Set<String> childHandles = handledTriggers.getOrDefault(child.name(), Set.of());
            if (!childHandles.contains(parentTransition.trigger())) {
                HierarchicalTransitionSpec fallback = new HierarchicalTransitionSpec(
                        child.name(), parentTransition.to(), parentTransition.trigger());
                parentTransition.requires().forEach(fallback::requires);
                parentTransition.produces().forEach(fallback::produces);
                out.add(fallback);

                propagateBubbling(roots, fallback, handledTriggers, out);
            }
        }
    }

    private HierarchicalStateSpec findState(List<HierarchicalStateSpec> roots, String name) {
        for (HierarchicalStateSpec root : roots) {
            HierarchicalStateSpec found = findStateRecursive(root, name);
            if (found != null) return found;
        }
        return null;
    }

    private HierarchicalStateSpec findStateRecursive(HierarchicalStateSpec node, String name) {
        if (node.name().equals(name)) return node;
        for (HierarchicalStateSpec child : node.children()) {
            HierarchicalStateSpec found = findStateRecursive(child, name);
            if (found != null) return found;
        }
        return null;
    }
}
